spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -1575,9 +1575,14 @@ struct ProductionVisitor : public production::Visitor {
         // Fields with stored elements should reserve the amount they will parse
         if ( auto* field = p->meta().field(); field && ! field->isTransient() ) {
             // Create a temporary so stateful size expressions are only evaluated once.
-            repeat = builder()->addTmp("repeat", repeat);
+            auto uint64 = builder()->qualifiedType(builder()->typeUnsignedInteger(64), hilti::Constness::Const);
+            repeat = builder()->addTmp("repeat", uint64, repeat);
 
-            builder()->addExpression(builder()->memberCall(destination(), "reserve", {repeat}));
+            // Limit the number of reserved elements just in case
+            auto max_reserve = builder()->integer(1U << 18);
```

Comment:
The general idea of this optimization is that we amortize some of the initial resizes when we start with capacity zero and repeatedly double it whenever elements beyond it are added; once we have some capacity these resizes become rarer.

With that I think it would make sense to pick a much, much smaller value for `max_reserve`, e.g., `512` or less instead of `262144`. This also helps when the vector parsing is stopped via `stop` as we do not allocated "too much" then. Additionally this should also be more conservative if the vector contains some bigger type.

```suggestion
            // Cap the number of reserved elements to avoid allocating
            // excessive much memory. The vector parsing might be stopped with
            // `stop` in which case we would not store the declared repeat
            // count anyway.
            auto max_reserve = builder()->integer(512);
            auto num_elements = builder()->min(max_reserve, repeat);
```

Reply:
Sure, seems mostly fine, since we should probably be conservative

It's no longer as big of a speedup (in fact no noticeable speedup for vectors with elements of single bytes), but maybe it would be for large elements. Regardless probably worth it to avoid OOM.

I don't see much utility in saying you'll parse `X` elements then planning on stopping early in some cases, though. Seems like there are better ways in spicy of doing that which won't reserve.